README
GOOMBAServer                                                                   README
|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

Legal Stuff
-----------

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

  See the file, COPYING, for the complete terms of the GPL.


Before You Begin
----------------

If you have absolutely no Unix experience, you may want to ask around
for someone with a little bit of Unix knowledge to help you through 
this installation.  Every attempt has been made to make it as simple
as possible, but since the software is quite involved and relies on a
number of different components, it is not realistic to assume it will
go smoothly on all systems.  You will probably need someone around who
knows the particulars of the destination system well.

If this is your first exposure to GOOMBAServer, please be aware that GOOMBAServer Version 
1.1 is quickly going to be superceded by Version 3.0.  There are
dramatic improvements.  Please see http://www.lerdorf.on.ca/GOOMBAServer3/
for more information, and if at all possible I suggest you start out
with GOOMBAServer3 instead of proceeding with GOOMBAServer2 today and then being forced
to migrate in a couple of months.

Windows NT Version
------------------

The code compiles almost cleanly using gnu-win32 gcc from Cygnus.
See http://www.cygnus.com/gnu-win32/

A pre-compiled Windows95/NT binary is available in the file archive at
http://GOOMBAServer.iquest.net/

You will need the Cygnus GCC DLL file available at:
ftp://ftp.cygnus.com/pub/gnu-win32/latest/cygwin.dll
Install this dll somewhere on your path.

You will also find a couple of MSVC 5.0 compiled binaries at
http://www.caraveo.com/GOOMBAServer/
These may or may not function better for you.

GOOMBAServer Version 3.0 will have much better Windows support, so if you are
interested in running GOOMBAServer on Windows, I suggest you start with version
3.0 available from http://www.lerdorf.on.ca/GOOMBAServer3/


Things You Need To Know Before Installing
-----------------------------------------

- Can you run both GET and POST method cgi programs on your server?

  If not, you can not use this package.  On many public ISP's CGI
  programs are either disallowed or severely restricted.  If this is
  the case on your system, talk to your system administrator and ask
  him/her to have a look at this package and see if they will install
  it for you.

- If you have mSQL installed on your system, you need to know the
  base directory of this installation.

- If you are going to be using this with sybase, you must have sybase db
  library and include files available. If you are not using this on DEC OSF1,
  remove the -ldnet_stub from src/Makefile, after running install.

- If you have Postgres95 installed on your system, you need to know the
  base directory of this installation.

- If you have mysql installed on your system, you need to know the
  base directory of this installation.

- You need an ANSI-C compatible C compiler (gcc is fine).

- If you are going to be using this with mSQL, you need to install
  mSQL before you install GOOMBAServer.

- If you are going to be using this with Postgres95, you need to
  install at least the libpq.a library from the Postgres95 
  distribution before installing GOOMBAServer.

- If you are going to be using gdbm, install gdbm before installing
  GOOMBAServer.

- If you are going to be using GD, install GD before installing GOOMBAServer.

- If you are installing the Apache module version, you will need to 
  know the Apache src code directory location.


FI Notes
---------

[This is more of a historical sub-note now.  GOOMBAServer and FI are one and
 the same today.]
 
FI is a subset of the GOOMBAServer distribution.  Throughout the documentation
GOOMBAServer and FI will be used interchangeably or together.  They refer to
the same thing.  On my system here the difference between FI and GOOMBAServer
is that the FI binary is compiled without the access logging and
access restriction code thus producing a faster parser for doing simple
database-enabled web pages.


Installation Instructions
-------------------------

If you do not have gcc installed on your system, make sure you turn 
on any compiler switches you need to make your compiler as ANSI 
compliant as possible.  The package is known to compile without
problems on the following systems.  If you manage to compile the
package on a system not listed here, please send me some e-mail
at: rasmus@lerdorf.on.ca with the name of your OS and the version 
number and any other details I should know.

Solaris 2.4
Solaris 2.5[.1]
Solaris 2.6
SunOS 4.1.2
SunOS 4.1.3
SunOS 4.1.4
Linux 1.3.x
Linux 1.1.x
Linux 2.1.x
BSDi 1.1
OSF1 (disable MMAP in config.h after running install)
FreeBSD (all versions)
NetBSD 1.2
NetBSD 1.2.1
Irix
Ultrix
HP-UX

Step 1.

  Run the install program: ./install

  You will be asked a number of questions.  If you do not understand
  what is being asked, simply hit return.  The default choice should
  be safe on most systems.

Step 2.

  Go into the src directory: cd src

  Have a look at the GOOMBAServer.h file.  There are a number of compile-time
  options that can be set here.

Step 3.

  type: make

  This will create the actual executable program file named GOOMBAServer.cgi
  by default.  (If you are NOT installing the Apache module version)

Step 4. (If you are NOT installing the Apache module version)
  
  Copy the GOOMBAServer.cgi binary to your system's cgi-bin directory.  If you
  do not have access to do this and wish to install it in your own
  personal directory, you may do so, but you should set the setuid
  bit on the executable with: chmod u+s /path/GOOMBAServer.cgi
  Setting the setuid bit is not crucial.  The benefit is that any files
  created by GOOMBAServer will be owned by you.  This means that you can edit
  and delete such files directly.  Without the setuid bit set these
  files will be owned by the httpd user id.

Step 4. (if you are installing the Apache module version)

  Change to your Apache src directory where the mod_GOOMBAServer.c ,mod_GOOMBAServer.h
  and libGOOMBAServer.a files should have been copied to. If they weren't which 
  usually happens because of permission problems, copy these three files 
  there manually. Edit your Apache Configuration file and add the 
  EXTRA_LIBS line which was produced at the end of Step 3. And add:

    Module GOOMBAServer_module mod_GOOMBAServer.o

  to the very end of the file. Then type: ./Configure and then 'make' to 
  rebuild your Apache httpd binary. Install this binary.

  Next you need to edit your Apache conf/srm.conf file and add a line like:

    AddType application/x-httpd-GOOMBAServer .phtml

  This defines a new MIME, application/x-httpd-GOOMBAServer, which will trigger the
  GOOMBAServer module to parse any file ending with the .phtml extension. You can pick
  any extension you like for this.

  Now you are ready to restart your httpd server. See the Apache Module
  Notes for more details on configuring the GOOMBAServer Module.

Step 5.

  Copy the doc/doc.html file to somewhere in your web document tree
  and read it using your favourite browser.  This file is the overall
  documentation for the package.


Testing the software
--------------------

  Once installed you can test to see if your executable works by 
  entering a URL similar to the following in your browser:

  http://your.site.domain/cgi-bin/GOOMBAServer.cgi

  This should show you a page which contains the version number along
  with various other useful information. 

  To test the Apache module version, create any file with a .phtml
  extension and put a tag like: <?GOOMBAServerinfo()> in the file and see if
  it gets parsed.


Using the software
------------------

  To actually use the software on an existing HTML file, you can 
  simply append the path to your file to the above URL.  ie.

  http://your.site.domain/cgi-bin/GOOMBAServer.cgi/path/file.html

  There are ways to automatically run your files through GOOMBAServer.  These
  methods are discussed in detail in the documentation.  

  Have a look in the examples directory for some examples of web pages
  developed using GOOMBAServer.

  Documentation, such that it is, can be found in the doc directory.


Troubleshooting and Mailing Lists
---------------------------------

  There is a GOOMBAServer mailing-list available for general discussion about
  GOOMBAServer.  If you are having difficulties, this is probably
  the best place to ask.  To sign up, send an empty message to:

  GOOMBAServer-list-subscribe@GOOMBAServer.iquest.net

  Please make sure you have read through doc/doc.html before asking
  questions on the mailing list. 

  Online documentation, plus the latest version of the software can
  be found at http://GOOMBAServer.iquest.net/

  There is a low-traffic mailing list which only contains announcements.
  To subscribe to it, send an empty message to:

  GOOMBAServer-announce-subscribe@GOOMBAServer.iquest.net

  All announcements will also be sent to the GOOMBAServer-list@GOOMBAServer.iquest.net mailing
  list, so you need not sign up to both.

  If you are interested in helping out with the development of GOOMBAServer,
  you can sign up to the GOOMBAServer-dev@GOOMBAServer.iquest.net list by sending an empty
  message to:

  GOOMBAServer-dev-subscribe@GOOMBAServer.iquest.net

  This list is quite technical and concentrates on code development.
  You will not be popular if you post general beginner questions here.
  There are also quite a few CVS commit log messages posted here.
  Especially during heavy hacking periods.


Sending in Bug Reports 
----------------------

  I appreciate receiving well-versed bug reports.  If they include a
  possible fix for the problem, I become quite estatic.  If you think
  you have found a bug and you can reproduce it, try to simplify the
  script that illustrates the bug to the least number of lines of
  code as possible.  Try compiling the CGI version of the package
  and pass it through that.  ie. ./GOOMBAServer.cgi filename.phtml
  
  If you are using the Apache module version and you get a 
  "Document contains no data" error, it is likely that the web server

ChangeLog
GOOMBAServer CHANGE LOG                                                    ChangeLog
|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
November 12 Version 1.1
- Add Ora_Bind() function <dizzy@glas.net>
- Add GetAllHeaders() function <ssb@guardian.no>
- Fix security holes in examples/{m,my}log.html <ssb@guardian.no>
- Add FPassThru function <ssb@guardian.no>
- Fix compile problem in snmp support <ssb@guardian.no>
- Fix syslog support in configure <ssb@guardian.no>
- Fix some Oracle bugs, add support for NULL columns <mitch@spiralmedia.com>
  and <ssb@guardian.no>
- Add support for persistent Oracle connections <mitch@spiralmedia.com>
- Fix array count/unset interactive bug
- Add RSort() and ARSort() functions to sort in descending order
- CGI version ~user bug fix
- Add GOOMBAServerLastModified Apache directive <pvanhaes@be.oracle.com>
- Fix typo in solid.c
- Fix \ bug in Sprintf()
- Add FilePro support <chadr@brttech.com>
- Add GOOMBAServerAutoPrependFile/GOOMBAServerAutoAppendFile Apache directives <ssb@guardian.no>
- Add StrTR() function <miko@ns.dcit.cz>
- Add initial Illustra support <simon@oyster.co.uk>
- Make isset() and unset() work correctly for array elements
- Fix NULL blob bug in Solid code <sopwith@redhat.com>
- Sybase connections were not being closed <scbiondi@webdesigngroup.com>
- UrlEncode shouldn't be encoding '/' - fixed
- Fix blank-line handling in command line -q mode <tbrown@baremetal.com>
- Fix elseif parser bug
- Fix possible 64-bit architecture alignment problem <doussot@via.ecp.fr>
- Push null string onto stack if GetEnv fails <eden@queer.org.au>
- Add shl() and shr() bit shifting functions <elanda@comstar.net>
- Fix buffer underrun in Exec() code <scott@ceweekly.com>
- Fix Oracle DateType field length bug <OF@netuse.de>
- Reverse dbm content-size fix from 1.1b12 for GDBM <W.Janssen@uci.kun.nl>
- Get rid of explicit calls to exit() from Sybase driver
- Add sybsql_checkconnect() function
- Add GDBM_FIX #define to GOOMBAServer.h
- Move CVS repostitory to inet-dev.bellglobal.com
- Move GOOMBAServer home web site to GOOMBAServer.iquest.net
- Fixed SAFE_MODE typos in head.c and var.c
- Change mailing list instructions to point to GOOMBAServer-list@GOOMBAServer.iquest.net
- A post'ed "0" should be treated as an integer, not a string - fixed
- Date("i") should print " 1" not "01" - fixed
- Fix critical bug in the internal FixFilename() function 
  <Meno.Abels@Technical.Adviser.com>
- Fix Oracle example in documentation <m.ramsch@ieee.org> 
- Fix dbm logging + DOCUMENT_ROOT bug
- Add first cut at ODBC support <brian@tool.net>
- Fix mysql_listfields() core dump <gareth@marvin.netcom.co.uk>
- Add Ceil() and Floor() functions
- Fix $req_* request variables with -'s in them (map to _)
- Add Adabas D support <kara@bibo.met.fu-berlin.de>
- Fixed blob handling for Adabas <kara@bibo.met.fu-berlin.de>
- Change QuoteMeta to escape ()'s
- Make Parser understand <?...?> syntax for forward compatibility with GOOMBAServer3
- Fix md5()-Apache 1.3 compatibility problem
- CVS repository moved to www.lerdorf.on.ca

June 16 Version 1.1b12
- Fix spinning httpd problem <terrym@iquest.net>
- Was munmap'ing 1 too few bytes - fixed
- Add array regression tests <fick@toadfish.fgm.com>
- Don't free returned pointer from getlogin() <kalkin@Ix.netcom.com>
- Fix core dump problems on some platforms caused by calling Exit from within
  a function in an included file
- Optional conversion of postgress field names to lower case in pg_result() 
  function (See GOOMBAServer_PG_LOWER in GOOMBAServer.h)
- Array assignment bug fixes
- Fix Include from inside a user-defined function
- Fix Set/Unset/Set count bug
- Have mysql interface check both getlogin and cuserid to find the current
  user id.
- Added ClearStack() hack/workaround function
- Fixed potential mmap leak
- ++ and -- on first 2 elements of an array were broken in 1.1b11 - fixed
- Fix bug associated with sorting an array with deleted items
- Fix command-line GOOMBAServer.cgi argv bug
- Change UrlEncode to encode a space with a +
- Fix PATH_DIR bug introduced in b11's security fix code
- Add Flush() command to flush output buffer for both module and CGI version
- Make System() call flush after each line of output
- Prepend owner uid to http auth realm string in safe mode
- Add $req_* variables for accessing request header values
- Fix SAFE_MODE Exec() calls to not cut off command line arguments
- Make pclose() return the status of the executed command
- Fix StrongHold related core dump
- Fix core dump in count() function when called with a bogus array
- Fix Eval() stack problem
- Fix Safe Mode problems related to dealing with non-existant files
- Fixed compiler warnings
- Fixed Solid_Close() bug
- Postgres install patch <jafo@tummy.com>
- Add Oracle install/Makefile support <pvanhaes@be.oracle.com>
- Add LogAs() function <eden@queer.org.au>
- Add NOLOGSUCCESSIVE directive to GOOMBAServer.h <eden@queer.org.au>
- Unset single element array bug fix <edell@EECS.Berkeley.EDU>
- Add regression test for the above
- Fix MySQL username confusion
- Add Stronghold compile option to install program
- Fix global variable scope bug
- Solid BLOB field fix <sopwith@redhat.com>
- Change GOOMBAServer* Apache directives to be accessible from .htaccess when 
  AllowOverride Options is set as opposed to FileInfo
- Add GOOMBAServerEngine Apache directive that is only callable from *.conf

src/acc.c
/***[acc.c]*******************************************************[TAB=4]****\
*                                                                            *
* GOOMBAServer                                                               *
*                                                                            *
* Copyright 2021,2022 GoombaProgrammer & Computa.me                          *
*                                                                            *
*  This program is free software; you can redistribute it and/or modify      *
*  it under the terms of the GNU General Public License as published by      *
*  the Free Software Foundation; either version 2 of the License, or         *
*  (at your option) any later version.                                       *
*                                                                            *
*  This program is distributed in the hope that it will be useful,           *
*  but WITHOUT ANY WARRANTY; without even the implied warranty of            *
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
*  GNU General Public License for more details.                              *
*                                                                            *
*  You should have received a copy of the GNU General Public License         *
*  along with this program; if not, write to the Free Software               *
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.                 *
*                                                                            *
\****************************************************************************/
#include <GOOMBAServer.h>
#include <stdio.h>
#include <stdlib.h>
#if HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <string.h>
#if HAVE_CRYPT_H
#include <crypt.h>
#endif
#include <ctype.h>
#include <errno.h>
#include <parse.h>
#include <regexpr.h>
#if APACHE
#include "http_protocol.h"
#endif

static char *RemoteHostName=NULL;
static char *EmailAddr=NULL;
static char *RefDoc=NULL;
static char *Browser=NULL;
#if defined(LOGGING) || defined(MSQLLOGGING)
static int Logging=1;
#endif
static int ShowInfo=1;


static AccessInfo *top=NULL;
static char *AccessDir=NULL;

#if APACHE
void GOOMBAServer_init_acc(GOOMBAServer_module_conf *conf) {
#else
void GOOMBAServer_init_acc(void) {
#endif
	RemoteHostName=NULL;
	EmailAddr=NULL;
	RefDoc=NULL;
	Browser=NULL;
#if defined(LOGGING) || defined(MSQLLOGGING)
#if APACHE
	Logging=conf->Logging;
#else
	Logging=1;
#endif
#endif
#if APACHE
	ShowInfo=conf->ShowInfo;
#else
	ShowInfo=1;
#endif
	top=NULL;
#if APACHE
	if(conf->AccessDir) AccessDir = estrdup(0,conf->AccessDir);
#if ACCESS_CONTROL
	else AccessDir = estrdup(0,ACCESS_DIR);
#else
	else AccessDir = NULL;
#endif
#else
#if ACCESS_CONTROL
	AccessDir = estrdup(0,ACCESS_DIR);
#else
	AccessDir = NULL;
#endif
#endif
}

#if ACCESS_CONTROL

AccessInfo *StrtoAccess(char *str) {
	AccessInfo *new, *last;
	char *s, *password=NULL;
	int state;
	int def=0, mode=0;

	while(top) {
		new = top->next;
		top = new;
	}
	last = NULL;
	top = NULL;
	s = strtok(str,"\033");
	state=0;
	while(s) {
		if(state==1 && *s>'Z') state=3;
		switch(state) {
		case 0: /* Get default mode */
			if(!strcmp(s,"pri")) def=0;
			else def=1; 
			state=1;
			break;
		case 1: /* Get mode bits */
			mode=0;
			if(strchr(s,'A')) mode |= 2;
			if(strchr(s,'B')) mode |= 4;
			if(strchr(s,'L')) mode |= 32;
			if(strchr(s,'E')) mode |= 1;
			if(strchr(s,'N')) mode |= 16;
			if(strchr(s,'S')) mode |= 64;
			if(strchr(s,'H')) mode |= 128;
			if(strchr(s,'P')) { 
				mode |= 8; 
				state=2; 
			}
			if(state==1) state=3;
			break;
		case 2: /* Get password */
			if(s && *s) {
				password = (char *) estrdup(0,s);
			} else {
				password = emalloc(0,sizeof(char));
				*password='\0';
			}
			state = 3;
			break;
		case 3: /* Get pattern */
			new = emalloc(0,sizeof(AccessInfo));
			new->password = password;
			new->mode = mode;
			new->def = def;
			new->next = NULL;	
			if(!strncmp(s,"dom",3)) new->type = 0;
			else if(!strncmp(s,"mai",3)) new->type = 1;
			else if(!strncmp(s,"ref",3)) new->type = 2;
			else if(!strncmp(s,"bro",3)) new->type = 3;
			else new->type = 0;
			new->patt = (char *)estrdup(0,&s[3]);
			StripSlashes(new->patt);
			if(!top) {
				top=new;
			} else {
				last->next = new;
			}
			last = new;
			state=1;
			break;
		}
		s = strtok(NULL,"\033");
	}
	return(top);
}

void AddRule(char *db) {
	VarTree *var;
	char temp[1024];
	char temp2[1024];
	char *s, *t;
	int ret;

	var = GetVar("addrule",NULL,0);
	if(!var) return;
	_dbmClose(db);
	ret = _dbmOpen(db,"w");
	if(ret) return;
	if(!strcmp(var->strval,"Default")) strcpy(temp,"cfg-accessALL");
	else sprintf(temp,"cfg-access-%s",var->strval);
	s = _dbmFetch(db,temp);	
	temp2[0] = '\0';
	if(s) {
		strcpy(temp2,s);
	} else {
		strcpy(temp2,"pub");
	}
	s = _dbmFetch(db,"cfg-accessALL");
	if(s) {
		strcat(temp2,"\033");
		t = strchr(s+4,'\033');
		if(t) {
			t = strchr(t+1,'\033');	
			if(t) *t='\0';
		}
		strcat(temp2,s+4);
	}
	_dbmReplace(db,temp,temp2);
	_dbmClose(db);
	_dbmOpen(db,"r");
}
		
void AddFile(char *db, char *path) {
	char temp[1024];
	char *pi, *s;
	int ret;

#if APACHE
	pi = GOOMBAServer_rqst->uri;
#else
	pi = getenv("PATH_INFO");
#endif
	if(!pi) return;
	sprintf(temp,"cfg-access-%s",pi);
	s = _dbmFetch(db,temp);
	if(s && strlen(s)>2) return;
	_dbmClose(db);
	ret = _dbmOpen(db,"w");
	if(ret) return;

	s = _dbmFetch(db,"cfg-accessALL");
	if(s) _dbmInsert(db,temp,s);	
	else _dbmInsert(db,temp,"pub\033L\033dom.*");
	_dbmClose(db);
	_dbmOpen(db,"r");
}

void ChkPostVars(char *db) {
	VarTree *var;
	int ret, op=0;

	var = GetVar("cfg-email-URL",NULL,0);
	if(var) {
		_dbmClose(db);
		ret = _dbmOpen(db,"w");
		if(ret) return;
		op = 1;
		_dbmReplace(db,"cfg-email-URL",var->strval);
		Push(var->strval,STRING);
		SetVar("cfg-email-URL",0,0);
	}
	var = GetVar("cfg-ban-URL",NULL,0);
	if(var) {
		if(!op) {
			_dbmClose(db);
			ret = _dbmOpen(db,"w");
			if(ret) return;
			op = 1;
		}
		_dbmReplace(db,"cfg-ban-URL",var->strval);
		Push(var->strval,STRING);
		SetVar("cfg-ban-URL",0,0);
	}
	var = GetVar("cfg-passwd-URL",NULL,0);
	if(var) {
		if(!op) {
			_dbmClose(db);
			ret = _dbmOpen(db,"w");
			if(ret) return;
			op = 1;
		}
		_dbmReplace(db,"cfg-passwd-URL",var->strval);
		Push(var->strval,STRING);
		SetVar("cfg-passwd-URL",0,0);
	}
	var = GetVar("chg-passwd",NULL,0);
	if(var) {
		if(!op) {
			_dbmClose(db);
			ret = _dbmOpen(db,"w");
			if(ret) return;
			op = 1;
		}
#if HAVE_CRYPT
		_dbmReplace(db,"cfg-passwd",(char *)crypt(var->strval,"xy"));
#else
		_dbmReplace(db,"cfg-passwd",var->strval);
#endif
		Push(var->strval,STRING);
		SetVar("cfg-passwd",0,0);
	}
	if(op) {
		_dbmClose(db);
		_dbmOpen(db,"r");
	}
}

void PostToAccessStr(char *db) {
	VarTree *var, *del=NULL;
	char ind[8];
	char temp[1024];
	static char temp2[512];
	int ret;
	int i, count;
	char small[16];

	var = GetVar("file",NULL,0);
	if(!var) return;
	_dbmClose(db);
	ret = _dbmOpen(db,"w");
	if(ret) return;
	if(!strcmp(var->strval,"Default")) strcpy(temp,"cfg-accessALL");
	else sprintf(temp,"cfg-access-%s",var->strval);

	var = GetVar("def",NULL,0);
	if(!var) return;
	if(!strcmp(var->strval,"public")) strcpy(temp2,"pub\033");
	else strcpy(temp2,"pri\033");

	var = GetVar("mode",NULL,0);
	count = var->count;
	for(i=0;i<count;i++) {
		sprintf(small,"del%d",i);
		del = GetVar(small,NULL,0);
		if(del && strchr(del->strval,'n')) continue;
		if(i>0) strcat(temp2,"\033");
		sprintf(ind,"%d",i);
		var = GetVar("mode",ind,1);
		sprintf(small,"%c",*(var->strval));
		strcat(temp2,small);
		if(*(var->strval) == 'P') {
			var = GetVar("password",ind,1);
			strcat(temp2,"\033");
			if(var && *(var->strval)) {
				strcat(temp2,var->strval);
			} else {
				strcat(temp2," ");
			}
		}
		var = GetVar("type",ind,1);
		if(var && var->strval) { 
			StripSlashes(var->strval);
			switch((var->strval)[0]) {
				case 'D':
					strcat(temp2,"\033dom");
					break;
				case 'E':
					strcat(temp2,"\033mai");
					break;
				case 'R':
					strcat(temp2,"\033ref");
					break;
				case 'B':
					strcat(temp2,"\033bro");
					break;
			}
		}
		var = GetVar("patt",ind,1);
		if(var && var->strval) {
			StripSlashes(var->strval);
			strcat(temp2,var->strval);
		}
	}
	if(strlen(temp2)<5) {
		_dbmDelete(db,temp);
	} else {
		_dbmReplace(db,temp,temp2);
	}
	_dbmClose(db);
	_dbmOpen(db,"r");
}

int CheckAccess(char *filename, long uid) {
	VarTree *var;
	char *s, *ss, *cp;
	char db[512], temp[512];
	AccessInfo *actop, *ac;
	struct re_pattern_buffer exp;
	struct re_registers regs;
	char fastmap[256];
	struct stat sb;
	int ret, allow=0;
	static char *email_URL=NULL, *passwd_URL=NULL, *ban_URL=NULL;
	int es;

	exp.allocated = 0;
	exp.buffer = 0;
	exp.translate = NULL;
	exp.fastmap = fastmap;

	if(stat(AccessDir,&sb)==-1) {
		if(mkdir(AccessDir,0755)==-1) {
			Error("Trying to create access directory [%s]: %d [%s]",AccessDir,errno,strerror(errno));
			return(0);
		}
	}
	sprintf(db,"%s/%ld-cfg",AccessDir,uid);

	es = ErrorPrintState(0);
	ret = _dbmOpen(db,"r");
	ErrorPrintState(es);
	if(ret) return(0);

	sprintf(temp,"cfg-access-%s",filename);	
	s = _dbmFetch(db,temp);
	if(!s || (s && !*s)) {
		s = _dbmFetch(db,"cfg-accessALL");
		if(!s || (s && !*s)) {
#if DEBUG
			Debug("Unable to read config string for this file\n");
#endif
			_dbmClose(db);
			return(0);
		}
	}
	actop = StrtoAccess(s);
	ac = actop;
	if(!ac) return(0);

	s = _dbmFetch(db,"cfg-email-URL");
	if(s) email_URL = (char *) estrdup(0,s);
	s = _dbmFetch(db,"cfg-passwd-URL");
	if(s) passwd_URL = (char *) estrdup(0,s);
	s = _dbmFetch(db,"cfg-ban-URL");
	if(s) ban_URL = (char *) estrdup(0,s);
	_dbmClose(db);

	while(ac) {
		cp = GOOMBAServer_re_compile_pattern(ac->patt,strlen(ac->patt),&exp);
		if(cp) {
			Error("Regular Expression error in rule: %s",cp);
			continue;	
		}
		GOOMBAServer_re_compile_fastmap(&exp);
		switch(ac->type) {
		case 0: /* domain */
			ss = getremotehostname();
			if(ss) {
				s = (char *) estrdup(0,ss);
			} else s=NULL;
			_strtolower(s);
			break;
		case 1: /* e-mail address */
			ss = getemailaddr();
			if(ss) {
				s = (char *) estrdup(0,ss);
			} else s=NULL;
			break;
		case 2: /* referring doc */
			s = getrefdoc();
			break;
		case 3: /* browser */
			s = getbrowser();
			break;
		}
		if(!s) { ac=ac->next; continue; }
		ret = GOOMBAServer_re_match(&exp,s,strlen(s),0,&regs);
		if(ret<1) { 
			ac=ac->next; 
			s=NULL;
			continue; 
		}
		switch(ac->mode) {
		case 1: /* E-Mail */
			if(!getemailaddr()) ShowEmailPage(email_URL);
			break;
		case 2: /* Allow */
			allow++;
			break;
		case 4: /* Ban */
			allow--;
			break;
		case 8: /* Password */
			var = GetVar("PASSWORD",NULL,0);
			if(!var || (var && strcmp(var->strval,ac->password))) ShowPasswordPage(passwd_URL);
			break;	
		case 16: /* NoLogging */
#if defined(LOGGING) || defined(MSQLLOGGING)
			Logging=0;
#endif
			break;
		case 32: /* Logging */
#if defined(LOGGING) || defined(MSQLLOGGING)
			Logging=1;
#endif
			break;
		case 64: /* Info */
			ShowInfo=1;
			break;
		case 128: /* NoInfo */
			ShowInfo=0;
			break;
		}
		ac = ac->next;
		s=NULL;
	}	
	if(actop->def==0) allow--;	
	if(allow<0) { ShowBanPage(ban_URL); return(-1); }
	return(0);
}

void ShowBanPage(char *url) {
	if(url && strlen(url) > 5) {
		GOOMBAServer_header(1,url);
		Exit(0);
		return;
	}
	GOOMBAServer_header(0,NULL);
	PUTS("<html><head><title>No Access</title></head>\n");
	PUTS("<body><h1>Sorry - You do not have access to this page</h1></body></html>\n");
	ShowInfo=0;
	Exit(0);
}

void ShowEmailPage(char *url) {
	VarTree *var;
	char *sn, *pi;
#ifdef VIRTUAL_PATH
	sn = VIRTUAL_PATH;
#else
#if APACHE
	sn = GOOMBAServer_rqst->uri;
#else
	sn = getenv("SCRIPT_NAME");
#endif
#endif
#if APACHE
	pi = sn;
#else
	pi = getenv("PATH_INFO");
#endif
	if(url && strlen(url) > 5) {
		GOOMBAServer_header(1,url);
		Exit(0);
		return;
	}
	GOOMBAServer_header(0,NULL);
	PUTS("<html><head><title>E-Mail Address Required</title></head>\n");
	PUTS("<body><h1>E-Mail Address Required</h1>Please provide your e-mail address to continue\n");
#if APACHE
	PUTS("<center><form action=\"");
	PUTS(sn);
	PUTS("\" method=\"POST\">\n");
#else
	printf("<center><form action=\"%s%s\" method=\"POST\">\n",sn,pi?pi:"");
#endif
	PUTS("<input type=\"text\" name=\"EMAIL_ADDR\">\n");
	var = GetVar("PASSWORD",NULL,0);
	if(var && *(var->strval)) {
#if APACHE
		PUTS("<input type=\"hidden\" name=\"PASSWORD\" value=\"");
		PUTS(var->strval);
		PUTS("\">");
#else
		printf("<input type=\"hidden\" name=\"PASSWORD\" value=\"%s\">\n",var->strval);
#endif
	}
	PUTS("<input type=\"submit\" value=\" Submit \"></form></center>\n");
	PUTS("</body></html>\n");
	ShowInfo=0;
	Exit(0);
	return;
}

void ShowPasswordPage(char *url) {
	VarTree *var;
	char *sn, *pi;

#ifdef VIRTUAL_PATH
	sn = VIRTUAL_PATH;
#else
#if APACHE
	sn = GOOMBAServer_rqst->uri;
#else
	sn = getenv("SCRIPT_NAME");
#endif
#endif
#if APACHE
	pi = sn;
#else
	pi = getenv("PATH_INFO");
#endif
	if(url && strlen(url) > 5) {
		GOOMBAServer_header(1,url);
		Exit(0);
		return;
	}
	GOOMBAServer_header(0,NULL);
	PUTS("<html><head><title>Password Required</title></head>\n");
	PUTS("<body><h1>Password Required</h1>Please enter the password:\n"); 
#if APACHE
	PUTS("<center><form action=\"");
	PUTS(sn);
	PUTS("\" method=\"POST\">\n");
#else
	printf("<center><form action=\"%s%s\" method=\"POST\">\n",sn,pi?pi:"");
#endif
	PUTS("<input type=\"password\" name=\"PASSWORD\">\n");
	var = GetVar("EMAIL_ADDR",NULL,0);
	if(var && *(var->strval)) {
#if APACHE
		PUTS("<input type=\"hidden\" name=\"EMAIL_ADDR\" value=\"");
		PUTS(var->strval);
		PUTS("\">\n");
#else
		printf("<input type=\"hidden\" name=\"EMAIL_ADDR\" value=\"%s\">\n",var->strval);
#endif
	}
	PUTS("<input type=\"submit\" value=\" Submit \"></form></center>\n");
	PUTS("</body></html>\n");
	ShowInfo=0;
	Exit(0);
	return;
}

#endif  /* ACCESS_CONTROL */

char *getremotehostname(void) {
	char *s;

	if(!RemoteHostName) {
#if APACHE
		s = table_get(GOOMBAServer_rqst->subprocess_env,"REMOTE_HOST");
#else
		s = getenv("REMOTE_HOST");
#endif
		if(s) {
			if(isdigit(*s)) s = (char *)_GetHostByAddr(s);
			RemoteHostName = s;
		} else {
#if APACHE
			s = table_get(GOOMBAServer_rqst->subprocess_env,"REMOTE_ADDR");
#else
			s = getenv("REMOTE_ADDR");
#endif
			if(s) {
				if(isdigit(*s)) s = (char *)_GetHostByAddr(s);
				RemoteHostName = s;
			}
		}
	}
	return(RemoteHostName);
}

char *getemailaddr(void) {
	static VarTree *var;

	if(!EmailAddr) {	
		var = GetVar("EMAIL_ADDR",NULL,0);
		if(!var) EmailAddr=NULL;
		else EmailAddr = var->strval;
	}
	return(EmailAddr);
}

char *getrefdoc(void) {
	if(!RefDoc) {
#if APACHE
		RefDoc = table_get(GOOMBAServer_rqst->subprocess_env,"HTTP_REFERER");
#else
		RefDoc = getenv("HTTP_REFERER");
#endif
	}
	return(RefDoc);
}	

char *getbrowser(void) {
	if(!Browser) {
#if APACHE
		Browser = table_get(GOOMBAServer_rqst->subprocess_env,"HTTP_USER_AGENT");
#else
		Browser = getenv("HTTP_USER_AGENT");
#endif
	}
	return(Browser);
}

int getlogging(void) {
#if defined(LOGGING) || defined(MSQLLOGGING)
	return(Logging);
#else
	return(0);
#endif
}

void setlogging(int val) {
#if defined(LOGGING) || defined(MSQLLOGGING)
	Logging = val;
#endif
}

void SetLogging(void) {
	Stack *s;

	s = Pop();
	if(!s) {
		Error("Stack error in setlogging()");
		return;
	}
	setlogging(s->intval);
}

int getshowinfo(void) {
	return(ShowInfo);
}

void setshowinfo(int val) {
	ShowInfo = val;
}

void SetShowInfo(void) {
	Stack *s;

	s = Pop();
	if(!s) {
		Error("Stack error in setshowinfo()");
		return;
	}
	setshowinfo(s->intval);
}

char *getaccdir(void) {
#if ACCESS_CONTROL
	if(AccessDir) return(AccessDir);
	else return(ACCESS_DIR);
#else
	return NULL;
#endif
}

void GetAccDir(void) {
#if ACCESS_CONTROL
	Push(getaccdir(),STRING);
#endif
}

